v2/brigadier/src/jobs.ts

```diff
@@ -28,6 +29,22 @@ export class Job implements Runnable {
   public logs(): Promise<string> {
     return Promise.resolve("skipped logs")
   }
+
+  public static pod(
```

Comment:
Tricky thing on the name here...

It's been a major goal of Brigade 2 to abstract everything away from Kubernetes so that k8s becomes merely an implementation detail. If it were ever to fall out of fashion, and some new orchestrator came en vogue, or the likes of Mesos experienced a renaissance, Brigade could live on with relatively few changes on the back end.

In light of this, we may want to reconsider `pod()` as a function name since it possibly tips its hand too much in terms of the underlying orchestrator.

Naming is hard.

Reply:
Hah, I was already planning to switch this helper to `container` - would that be okay?

Reply:
Maybe it depends on what "container" means here.

If it means "a container-_based_ job, that seems accurate.

But it could also mislead people into believing a job (singular) is implemented as container (singular), when in reality a job can be composed of multiple containers.

Add in my general reservations about revealing too much about the underlying implementation........ and then I have to back-pedal and acknowledge that the word container is already used in a bunch of other places. ::sigh::

Then another consideration would be that (if I understand correctly), Krustlet is enabling things that _aren't_ containers at all (namely web assembly modules) to run in k8s. If that works as advertised, I believe a job can conceivably be composed of those instead, without a single change to Brigade.

Smoke is coming out of my head as I grapple with whether "container" can simultaneously be accurate-enough _and_ not overly-specific about the underlying implementation.

But as I grapple with it, I'm coming to wonder if we need it...

What would `Job.container(...)` buy us that we don't already get from `new Job(...)`? In terms of the other two new static functions, I see that they address concerns over discoverability / context help, and make `Job` a one-stop shop, but I'm less clear on the benefits of `Job.container(...)`....... but I'm also not as well versed in common TS idioms for this sort of thing. 🤷‍♂️

Reply:
I was hesitant about proposing it and could be talked out if it.  The reasons I proposed it were essentially it's for consistency and explicitness.  Consistency in that it makes `Job.` a one-stop shop for _all_ the ways you can construct things to run; and explicitness in saying "although Job can produce other kinds of Runnable, this gives you the _particular kind_ which is executed by running a container."

And the trouble is... if "a Job runs a container" isn't true, then _how do we want to document Job?_  What do we tell people a Job is?  Is it "it runs an OCI image"?  But then you can say "but what if one day we do WASM stored in Bindle."  Is it "an external program, packaged in whatever way"?  But then what if we decide that Bindled WASM jobs need to be represented differently, and we need to contract the meaning of Job to be specifically OCI?

So... I'm wary of second guessing.  Today, a Job is basically a container (Krustlet notwithstanding).  I'm happy documenting it as such, and therefore I'm happy using the term in a factory method.  Tomorrow, maybe we'll need to reconsider,

But possibly that's a third benefit of the factory method.  Suppose we want to support Bindled WASM, and we find that requires different info from a container.  (Say, it needs a selector, and it can't have a `command`.)  Well, `Job.container` and `Job.bindle` make that difference clear and can have different signatures.  Whereas `new Job` will become ambiguous and might need optional parameters and runtime configuration errors.

Of course, the converse is that if we are happy to say "a Job today _is_ a container", then perhaps we should be happy to defer the explicitness argument, and stick to `new Job` until Job _does_ become ambiguous.  I dunno.  I'm thinking out loud here, trying to figure out how we document Job (and whether I can then turn that into a terse method name).

For myself, I like the explicitness and I like the "I can type `Job.` and find out what my options are" effect.  But I totally agree that it is close to redundant, and may be more specific than we want to be, and if you do object then I won't shed many tears about taking it out again.

Hope this all makes sense; happy to discuss on a call if face-to-face would be easier!

Reply:
Well now both our heads are smoking grappling with naming.

I really like your point supposing we did need to someday define different sorts of jobs differently-- like a WASM based job hypothetically requiring additional fields.

Based on that argument, consider me sold. Your approach leaves room for expansion without ambiguity.

Any thoughts on changing `container()` to `containerBased()`? Could that convey the manufacture of a container-_based_ job without opening the door to misinterpretation that a job is a (singular) container?

Reply:
For my money `containerBased` kind of muddies the water (though I see where you are coming from).  My interpretation of a job was that it ran a container (or commands in that container) _but_ you could stick sidecars on in case you wanted to proxy network calls or something - but the sidecars were ancillary not job-defining.  But if the secondary containers are expected to perform actual job-related work, then maybe we need to open the naming to that.

Reply:
> But if the secondary containers are expected to perform actual job-related work, then maybe we need to open the naming to that.

We're not opinionated about what they do, but the "rules" are the exit code of the primary container is the determinant of job success or failure. The exit codes of the others aren't ever checked. When the primary container exits, the whole job is considered done. One minute later (configurable), we delete the whole pod, because k8s is really unintelligent about this. k8s makes not actual distinction between your "primary" container and a sidecar, so when the primary exits, the sidecars stupidly could remain running indefinitely. There are several extensive threads about this defficiency in k8s. We circumvent all of that by torching the whole pod-- something we have the luxury of doing in Brigade 2 that we couldn't have done in 1.x-- because unlike 1.x, 2.0 aggregates logs and stores them somewhere. 1.x had to hang on to your pods for as long as you wanted to keep the logs (and even then you could still lose them).

I'm getting off track. Long story short is that the view of _a_ job as _a_ container is probably not quite right. It is, without a doubt, the common case, but jobs can be an assembly of multiple containers.